### PR TITLE
Chol incremental

### PR DESCRIPTION
**Pull request by Aneoshun**
Hi all, 

After my deceptive attempt to use the block-wise inversion, I decided to take a look on how BayesOpt does it. Conclusion, they are not using block-wise inversion, but incremental Cholesky decomposition (computing the additional line of the L matrix).
I should have thought about this early, but anyway, I implemented it and the results are quite good!
The computation of the kernel is 3 times faster in average (most of the time is used in the matrix allocations/resizings).
I also added a check that verifies that there were some changes in the samples or in the kernel_function between two calls of "compute", if not then, no need to re-compute the kernel (useful, when only adding a BL sample)

As soon as this branch is merged, I will update the benchmarks.

### CODE REVIEW

**Comment by jbmouret** [comment]
That's great! We could make the kernel computation even faster by
pre-allocating a big matrix. Do you think it could make sense or this would
not make any difference at the end?

-- JBM

**Comment by Aneoshun** [comment]
Not sure.
To be more precise, when I said "allocation/resize" I was talking about the whole process including computing the new data. The re-allocation is not really long, it's more filling the data in the new matrix which takes time. 
And from my experience, each time I tried to "re-implement" something that I thought Eigen is doing too slowly, I end with something several time slower.
Here is the measurements of the different parts (values in us):
```
resize kernel :3
filling kernel :399
computing LLT decomp :218
computing _alpha: 37
Time running first batch: 707us

conservative resize of kernel: 81
filling kernel: 5
conservative resize of matrixL: 74
filling  matrixL: 35
computing _alpha: 19
Time running increment: 260us

Time running no change: 1us

resize kernel :1
filling kernel :362
computing LLT decomp :215
computing _alpha: 19
Time running whole batch: 632us
```

I am not convinced we can easily do better.

**Comment by fedeallocati** [comment]
I've got a question. Why not extend the GP API to provide an explicit method to add new samples? Something like `add_sample(const Eigen::VectorXd& sample, const Eigen::VectorXd& obs)` and `add_bl_sample(const Eigen::VectorXd& sample)`. In this way, the `BOptimizer` can explicitly specify that this is a new sample (and straightforwardly compute just the new row of the L matrix, without doing checks to see if the rest of the samples changed), and the GP Optimization can use `compute(...)` like now, that always recompute the full matrix (also without doing any checks).
I think it's both cleaner and more efficient, as we don't have to check every time that the samples haven't changed

**Comment by Aneoshun** [comment]
I agree with you. 
The main reason I did not implemented this is because I wanted something transparent (only altering the "compute" and "_compute_kernel" from gp.hpp) and also because I wasn't sure about the result. 

What you suggest is doable, but requires to change a bit more in depth the lib. 
If someone wants to give a try, he is more than welcome.

**Comment by fedeallocati** [comment]
If there are no objections to this, I can try and do it

**Comment by jbmouret** [comment]
No objection.

Le mar. 10 mai 2016 15:49, Federico Allocati <notifications@github.com> a
écrit :

**Comment by costashatz** [comment]
No objection...

**Comment by fedeallocati** [comment]
The task is finished. I have added two new methods:
`void add_sample(const Eigen::VectorXd& sample, const Eigen::VectorXd& observation, double noise)`
`void add_bl_sample(const Eigen::VectorXd& bl_sample, double noise)`
This is the Incremental API, which uses the new incremental cholesky decomposition to add the samples.
The old `compute` function now works as a method for retraining the GP from scratch, without taking into account the previously added samples.
I also modified the method `update()` to be more explicity: `recompute(bool update_obs_mean)` on which the GP recomputes the full kernel (and if indicated so, the `obs_mean` matrix) with the current samples and observations, for the HP Optimizers to use. Also I fixed a bug in this method, on which we weren't recomputing the blacklisted kernel with the new hyperparameters.
Regarding the GP, I made a small improvement in the calculation of the `obs_mean` using Eigen's `colwise().mean()` instead. I noticed that there are many serial `for` loops that could be replaced by the parallel for, thus taking advantage of multicore, but a more in-depth analysis should be made, to consider if it is usefull in all cases. 
Finally, I modified the `tests_gp` to use the incremental API. Since now there is no way of doing nothing, I removed that test, and added a check with the `recompute` method.
Please check to see if I haven't made any mistake, and if you like the way I solved this!

**Comment by fedeallocati** [comment]
PS. Now there is something weird with this API and the `noise` parameter. The API would technically allow to have a different amount of noise for each sample, but then when calling `recompute()`, only the last noise is taken into consideration. Maybe this could be improved, or changed to have a unique noise independent of the samples.
I also did some code refactoring, and a little improvement when computing the full kernel: Instead of computing the kernel function for all i, j pairs, now we compute only for the lower part of the K matrix, and then copy from the lower to the upper part. I guess this should be more efficient than computing it twice for each pair

**Comment by jbmouret** [comment]
We could move the noise to the Params::gp (Params::gp::noise). What do you think?

**Comment by Aneoshun** [comment]
Thanks for your work. 
It seems to work very well. A good test would be to re-run the benchmark. I will do it as soon as this branch is merged.

Regarding the noise parameter, it's an interesting question. 
From my point of view, forcing the noise to the same level limits the abilities of the algorithms (we had a few ideas about multi-sources learning, based on different, learned, noise values).

We could treat the noise values as an additional attribute (like _observations) and accumulate the values into a vector, which allows afterward to recompute the kernel properly. 

What do you think?

File: src/limbo/bayes_opt/boptimizer.hpp [hunk 1]
```diff
@@ -35,18 +35,18 @@ namespace limbo {
         /**
         The classic Bayesian optimization algorithm.
 
-        \rst
+        \\rst
```
Comment (jbmouret) [inline src/limbo/bayes_opt/boptimizer.hpp]: Why this? It seems the documentation generation was working on this file! see: http://www.resibots.eu/limbo/api.html#_CPPv2N5limbo9bayes_opt10BOptimizerE

**Comment by jbmouret** [comment]
We can merge this branch and the benchmark branch.

For the noise, I would love a support for multiple noise, but I am not sure
this is as trivial as just storing the noise that correspond to each
sample. Is it?

In the GPML book, they say (p. 191): "observed function and derivatives may
often have different noise levels, which are incorporated by adding a
diagonal contribution with differing hyperparameters. Inference and
predictions are done as usual". This suggests that this is simple.

In that case, we could simply have a vector to store the noise values.

What do you think?

**Comment by Aneoshun** [comment]
From my point of view, it's as easy as that.

File: src/limbo/bayes_opt/boptimizer.hpp (hunk 1 above)
Comment (fedeallocati) [inline src/limbo/bayes_opt/boptimizer.hpp]: Ooops. I had an issue with my linux installation and had to retrieve both files with a recovery tool. It changed enters with literals \n and quotes with \" so I replaced them, but I didn't noticed this. I will fix them ASAP

**Comment by Aneoshun** [comment]
Hi, 
I change the noise into a vector. 
Currently this changes nothing because the boptimizer always uses the noise value in Params. However, the GP can now handle different values of noise.

Ready to merge?

File: src/limbo/bayes_opt/boptimizer.hpp [hunk 2]
```diff
@@ -91,7 +91,7 @@ namespace limbo {
                 this->_init(sfun, afun, reset);
 
                 if (!this->_observations.empty())
-                    _model.compute(this->_samples, this->_observations, Params::bayes_opt_boptimizer::noise(), this->_bl_samples);
+                    _model.compute(this->_samples, this->_observations, Eigen::VectorXd::Constant(this->_observations.size(), Params::bayes_opt_boptimizer::noise()), this->_bl_samples);
```
Comment (costashatz) [inline src/limbo/bayes_opt/boptimizer.hpp]: Maybe we could have a vector as a parameter? It makes more sense...

**Comment by Aneoshun** [comment]
Not really, because you do not know beforehand how many samples you will have. And, if the purpose is to assign the default value everywhere, no need to bother the users to define a vector.

**Comment by costashatz** [comment]
> Not really, because you do not know beforehand how many samples you will have. And, if the purpose is to assign the default value everywhere, no need to bother the users to define a vector.

Ohh yes.. You are right.. Too much in my head lately.. Sorry..

**Comment by costashatz** [comment]
> Hi, 
> I change the noise into a vector. 
> Currently this changes nothing because the boptimizer always uses the noise value in Params. However, the GP can now handle different values of noise.
> 
> Ready to merge?

**OK** for me...

**Comment by fedeallocati** [comment]
Let me check it today. Im moving to a new apartement right now, but I want
to check it before merging!

**Comment by jbmouret** [comment]
The proper way would be to add this in the optimized function (like dim_in
& dim_out). But I do not think if it is worth the trouble.

**Comment by Aneoshun** [comment]
Sure! There is no urgency. I was just wandering if something else needed to be done (I am working on the ARD kernel). 
Good luck with the moving.

**Comment by fedeallocati** [comment]
I have two points to ask:
- W.r.t. hp optimization, why does -1 is the reserved value for no
optimization? I think it means more sense for it to be 0 since iters % 0
would blow up and I think it is more natural.
- W.r.t. noise, does it make sense for blacklisted samples to have noise? I
remember when once J.B. gave me an example of the usefulness of having
different noise value per sample(when there is a sensor and a way to
measure noise on the measurement). In that case, a blacklist could be the
sensor giving a weird value, and there would be no noise tu measure. So I
think that any case when there is no observation, also makes sense to do
not have noise. On the other hand, if Im not wrong, I think that the noise
has the effect of adding some uncertainity on the blacklisted samples,
which can be useful.

Tell me what do you think of this points!

**Comment by Aneoshun** [comment]
Regarding the noise, your "On the other hand" option is absolutely right.
The noise can represent the inverse probability of that point to generate BL samples. 
It can be useful is the problem related to this samples is stochastic. 

Regarding the HP Opt, I think we should say 
```
if (Params::bayes_opt_boptimizer::hp_period() > 0
                        && this->_current_iteration % Params::bayes_opt_boptimizer::hp_period() == 0)
```
to avoid any problem and use any value <=0 to prevent optimization.

**Comment by costashatz** [comment]
> Regarding the HP Opt, I think we should say
> 
> if (Params::bayes_opt_boptimizer::hp_period() > 0
>                         && this->_current_iteration % Params::bayes_opt_boptimizer::hp_period() == 0)
> to avoid any problem and use any value <=0 to prevent optimization.

You're right!

**Comment by costashatz** [comment]
Should we merge?

**Comment by Aneoshun** [comment]
It seems so to me.

**Comment by costashatz** [comment]
**Let's merge...**

**Comment by costashatz** [comment]
As I was trying to fix the errors in the experimental things, I came to the fact that we need to augment also the GPParego model to have functions `add_sample`, `add_bl_sample` and accept a vector of noises. I am not sure how GPParego works, @jbmouret ?

Also, we should remove the `add_sample` and `add_bl_sample` functions from `BOptimizer` right?

**Comment by jbmouret** [comment]
Parego should compile now.

We should probably remove ``add_sample`` and ``add_bl_sample``. Is it used somewhere?

**Comment by Aneoshun** [comment]
It happens that the test_gp fails because of the check regarding the execution time:
```
Time running first batch: 380us

Time running increment: 345us

Time recomputing: 221us

Time running whole batch: 253us

../src/tests/test_gp.cpp:170: error: in "test_gp_bw_inversion": check time_full > time_increment has failed
../src/tests/test_gp.cpp:171: error: in "test_gp_bw_inversion": check time_recompute > time_increment has failed
```
Actually, I don't know why this sometimes happens, but it is perhaps better to remove these two checks. What do you think?

**Comment by jbmouret** [comment]
I think you wrote this test, right?

**Comment by Aneoshun** [comment]
Yes, I did and it works most of the time, but it depends on external factors that make it fail from time to time.

Therefore, two possibilities: remove the test, or perform replicates of the evaluation and compare statistics (to marginalize the external factors). 
While the second option is the most rigorous, I think that it's perhaps a bit overkilled.

The first part of this test compare the resulting L matrices to check if they are similar. I think that this test can be enough.

**Comment by costashatz** [comment]
> While the second option is the most rigorous, I think that it's perhaps a bit overkilled.

I think I prefer this way. I can do it if you like..

**Comment by Aneoshun** [comment]
Sure, go ahead if you want!

Envoyé de mon iPhone

**Comment by costashatz** [comment]
> We should probably remove add_sample and add_bl_sample. Is it used somewhere?

What will we do? I do not find any good usage of these functions as we have them now..

**Comment by costashatz** [comment]
**Are we ready to merge then?**

**Comment by jbmouret** [comment]
**OK for me**

**Comment by fedeallocati** [comment]
**Ok for me**

**Pull request merged**